### ISSUE TITLE
[Bug]: `fig.set_dpi()` and `figure.dpi` option in `matplotlibrc` file not working

### ISSUE THREAD

**Issue by BearBearCodes** [matplotlib/matplotlib#24644]
### Bug summary

- `fig.set_dpi(150)` only changes the dpi of plot shown in the Jupyter notebook but not in the saved figure
- `figure.dpi` in the `matplotlibrc` file does not work at all
- `plt.subplots(dpi=150)` works as expected and changes the dpi of both the plot shown in the Jupyter notebook as well as in the saved figure

Continuation of discussion in #11227

### Code for reproduction

```python
import matplotlib as mpl
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
fig.set_dpi(150)
ax.plot([1, 2], [1, 2])
print(fig.get_dpi())
print(mpl.rcParams["figure.dpi"])
print(mpl.rcParams["savefig.dpi"])
fig.savefig("test.png", dpi="figure")
plt.show()
```

### Actual outcome

Inside a Jupyter notebook running python 3.10.4 inside a conda environment (no `matplotlibrc` file):
```text
150
72.0
figure
```
![image](https://user-images.githubusercontent.com/64443273/205959252-8bd47647-ff7a-41e0-8d63-f45c974fa5d0.png)
Saved plot is at 72.0 dpi.

Inside a standalone Python script running the same python version inside a conda environment (no `matplotlibrc` file):
```text
150
100.0
figure
```
Saved plot is at 100.0 dpi.

### Expected outcome

- `fig.set_dpi(150)` would set the dpi to 150 for the figure shown in the Jupyter notebook + the figure shown in the matplotlib window that pops up when `plt.show()` is called as well as the figure saved.

- `figure.dpi` in the `matplotlibrc` file would change the default dpi of figures created (but it does not). From my original comment in [#11227](https://github.com/matplotlib/matplotlib/issues/11227#issuecomment-1338755358):
    > Moreover, I'm finding that on my machine, the `figure.dpi` rcParam is not respected after I have modified it in my `matplotlibrc` file (located at `~/.config/matplotlib/matplotlibrc`). Other changes to options like `axes.prop_cycle` in my `matplotlibrc` file are being reflected in the plots shown in the Jupyter notebook as well as in those saved via `fig.savefig()`, but changes to the `figure.dpi` option are not reflected in the plots shown/saved

- The default dpi should be 100.0 instead of 72.0 (only seems to happen inside Jupyter notebooks and not in standalone python scripts)

- (similar to point <span>#</span>1) `fig.savefig("test.png", dpi="figure")` should reflect the changed dpi that was set with `fig.set_dpi(150)`. Instead, it saves the figure at 72.0 dpi (when running the code in a Jupyter notebook) or 100.0 dpi (when running the code in a standalone Python script). Both the Jupyter notebook and Python script are running the same Python version in the same conda environment.

### Additional information

Previous discussion in #11227

### Operating system

WSL2 running Ubuntu 20.04.5 LTS

### Matplotlib Version

`3.5.2`

### Matplotlib Backend

Inside Jupyter notebook (same regardless whether or not `matplitlibrc` file exists): `module://matplotlib_inline.backend_inline`.Inside Python script with no `matplotlibrc` file: QtAgg

### Installation

None

**Comment by tacaswell** [comment]
Do you have a high-dpi screen?

**Comment by BearBearCodes** [comment]
Yes I sometimes use a high dpi screen but I've noticed this behaviour even when I'm only using my laptop screen (1920×1080) with no external displays connected.

**Comment by tacaswell** [comment]
The source of the problem is this:

 - figures know what their DPI is
 - this dpi is part of the transform stack and going from our various internal coordinate systems out to the final screen units (either pixel or points depending on the target format)
 - for high-dpi displays we double the resolution behind the scenes so that things look good on the screen [1]
 - to make sure that doubling does not propagate out we stash a copy of the initial dpi [2]
 - and use that cached value in the case of `dpi='figure'` [3]
 - however we do not update that cached value if the user changes it

Doing `fig, ax = plt.subplots(dpi=150)` works as expected (because we set it early).  `mpl.rcParams['figure.dpi'] = 150` should work, but you may have to update `matplotlib-inline` to get a version that does not change rcParams on import for it to work set via `matplotlibrc`.


[1] https://github.com/matplotlib/matplotlib/blob/b2e843405de323c9170b9da9f300dc6f372cde18/lib/matplotlib/backend_bases.py#L2117-L2123
[2] https://github.com/matplotlib/matplotlib/blob/b2e843405de323c9170b9da9f300dc6f372cde18/lib/matplotlib/backend_bases.py#L1673
[3] https://github.com/matplotlib/matplotlib/blob/b2e843405de323c9170b9da9f300dc6f372cde18/lib/matplotlib/backend_bases.py#L1673


So we need at least one more go-around to get this correct. I am going to see what it looks like to:
 - move `_original_dpi` into `Figure`
 - make public facing ways to change the dpi update this
 - make sure we _never_ use the public API internally

**Comment by tacaswell** [comment]
tagging as 3.7 as the amount of changes I am planning I expect to be bigger than what I would like to backport, but can be talked out of that (and probably should wait to see the code before jumping to conclusions anyway).

**Comment by QuLogic** [comment]
Oh, I thought I saw the `set_dpi` issue before, maybe I didn't fix it? I expect this would be fixed by my refactor of the HiDPI handling, but I've not touched that in some time.

**Comment by tacaswell** [comment]
As an update, I took at pass at this and failed.  @QuLogic has a plan to fix this at a lower level.

**Comment by jklymak** [comment]
Just to clarify this a bit, because I think its being muddied by hiDPI and inline issues:

### dpi set on creation:

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots(dpi=150)
ax.plot([1, 2], [1, 2])
plt.show()
fig.savefig('Boo.png', dpi='figure')
```

`Boo.png` has 150 dpi

### dpi set after creation:

```python
fig, ax = plt.subplots()
fig.set_dpi(150)
ax.plot([1, 2], [1, 2])
plt.show()
fig.savefig('Boo.png', dpi='figure')
```

`Boo.png` has the dpi of 100.  

No doubt because https://github.com/matplotlib/matplotlib/blob/a5f5af30abf09e62bd70c4dbbf4af921d73ab88a/lib/matplotlib/backend_bases.py#L2301 sets the printing dpi to `fig._original_dpi`.

**Comment by jklymak** [comment]
Note that if we just modify `set_dpi` to change `_original_dpi` that fixes the problem here.  

It doesn't fix the original issue #11227, which is somewhat orthogonal.

**Comment by jklymak** [comment]
So far as I can tell, `set_dpi` is completely untested.  I'd actually propose deprecating it, as there is not much reason to call it.

**Comment by tacaswell** [comment]
`Figure.set_dpi` works correctly if you have a figure open interactively, this is a bug we should fix.
